File: utils/filter_by_date.go
package utils

import (
	"rahul-rakshit/formatbankcsv/output"
	"time"
)

func FilterByDate(data [][]string, fromDate string, toDate string) ([][]string, error) {
	var filteredData [][]string
	for _, row := range data {
		err := validateDates(row, fromDate, toDate)
		if err != nil {
			return [][]string{}, err
		}

		if isHeaderRow(row) || (meetsFromDateBound(row, fromDate) && meetsToDateBound(row, toDate)) {
			filteredData = append(filteredData, row)
		}
	}

	return filteredData, nil
}

func validateDates(row []string, fromDate string, toDate string) error {
	layout := "2006-01-02"

	_, rowDateErr := time.Parse(layout, row[0])
	if rowDateErr != nil && row[0] != output.Header[0] {
		return rowDateErr
	}

	_, fromDateErr := time.Parse(layout, fromDate)
	if fromDateErr != nil && fromDate != "" {
		return fromDateErr
	}

	_, toDateErr := time.Parse(layout, fromDate)
	if toDateErr != nil && toDate != "" {
		return fromDateErr
	}

	return nil
}

func isHeaderRow(row []string) bool {
	return row[0] == output.Header[0]
}

func meetsFromDateBound(row []string, fromDateString string) bool {
	if fromDateString == "" {
		return true
	}

	layout := "2006-01-02"
	rowDate, _ := time.Parse(layout, row[0])
	fromDate, _ := time.Parse(layout, fromDateString)

	return rowDate.After(fromDate) || rowDate.Equal(fromDate)
}

func meetsToDateBound(row []string, toDateString string) bool {
	if toDateString == "" {
		return true
	}

	layout := "2006-01-02"
	rowDate, _ := time.Parse(layout, row[0])
	toDate, _ := time.Parse(layout, toDateString)

	return rowDate.Before(toDate) || rowDate.Equal(toDate)
}

File: utils/filterByDate.ts
import { Header } from "./constants.ts";

function validateDate(dateStr: string): boolean {
  if (dateStr === "" || dateStr === Header[0]) return true;
  const date = new Date(dateStr);
  return date instanceof Date && !isNaN(date.getTime());
}

function validateDates(row: string[], fromDate: string, toDate: string): void {
  const rowDate = row[0];
  
  if (!validateDate(rowDate) && rowDate !== Header[0]) {
    throw new Error(`Invalid date format: ${rowDate}`);
  }
  if (!validateDate(fromDate)) {
    throw new Error(`Invalid fromDate format: ${fromDate}`);
  }
  if (!validateDate(toDate)) {
    throw new Error(`Invalid toDate format: ${toDate}`);
  }
}

function isHeaderRow(row: string[]): boolean {
  return row[0] === Header[0];
}

function meetsFromDateBound(row: string[], fromDateString: string): boolean {
  if (fromDateString === "") {
    return true;
  }

  const rowDate = new Date(row[0]);
  const fromDate = new Date(fromDateString);

  return rowDate >= fromDate;
}

function meetsToDateBound(row: string[], toDateString: string): boolean {
  if (toDateString === "") {
    return true;
  }

  const rowDate = new Date(row[0]);
  const toDate = new Date(toDateString);

  return rowDate <= toDate;
}

export function filterByDate(
  data: string[][],
  fromDate: string,
  toDate: string
): string[][] {
  const filteredData: string[][] = [];

  for (const row of data) {
    try {
      validateDates(row, fromDate, toDate);
    } catch (error) {
      throw error;
    }

    if (isHeaderRow(row) || (meetsFromDateBound(row, fromDate) && meetsToDateBound(row, toDate))) {
      filteredData.push(row);
    }
  }

  return filteredData;
}
